extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java

```diff
@@ -124,6 +124,17 @@ AdditionalIndexedClassesBuildItem includeMongoCommandListener(MongoClientBuildTi
         return new AdditionalIndexedClassesBuildItem();
     }
 
+    @BuildStep
+    AdditionalIndexedClassesBuildItem includeMongoCommandMetricListener(
+            MongoClientBuildTimeConfig buildTimeConfig,
+            Optional<MetricsCapabilityBuildItem> metricsCapability) {
+        return metricsCapability
+                .filter(cap -> buildTimeConfig.metricsEnabled && cap.metricsSupported(MetricsFactory.MICROMETER))
+                .map(supported -> new AdditionalIndexedClassesBuildItem(
+                        MongoClientRecorder.getMicrometerCommandListenerClassName()))
+                .orElseGet(AdditionalIndexedClassesBuildItem::new);
```

Comment:
Any chance you could implement this with simple ifs (with early returns for edge cases)?

Because this is a bit hard to follow and I'm worried we will regret it when fixing a bug one early morning without coffee :).

Reply:
@gsmet pushforced the change